syfertext/tokenizer.py

```diff
@@ -175,18 +217,411 @@ def __call__(self, text: Union[String, str]):
+        return doc
+
+    def _tokenize(self, substring: str, token_meta: TokenMeta, doc: Doc) -> Doc:
+        """ Tokenize each substring formed after splitting affixes and processing 
+            exceptions. Returns Doc object.
+
+        Args:
+            substring: The substring to tokenize.
+            token_meta: The TokenMeta object of original substring
+                before splitting affixes and exceptions.
+            doc: Document object. 
+
+        Returns:    
+            doc: Document with all the TokenMeta objects of every token after splitting 
+                affixes and exceptions.
+        """
+
+        # Start position of substring in text to be tokenized.
+        pos = token_meta.start_pos
+
+        # If there is trailing space after the substring in text.
+        space_after = token_meta.space_after
+
+        # Get the remaining substring, it's start pos relative to original text,
+        # affixes containing list of TokenMeta for each type affix and
+        # list of TokenMeta of exceptions after splitting the affixes.
+        substring, pos, affixes, exception_tokens = self._split_affixes(
+            substring=substring, start_pos=pos
+        )
+
+        # Attach all the `TokenMeta` objects formed as result of splitting
+        # affixes in the doc container.
+        doc = self._attach_tokens(
+            doc=doc,
+            substring=substring,
+            start_pos=pos,
+            space_after=space_after,
+            affixes=affixes,
+            exception_tokens=exception_tokens,
+        )
 
         return doc
 
+    def _split_affixes(
+        self, substring: str, start_pos: int
+    ) -> Tuple[str, int, DefaultDict, List[TokenMeta]]:
+        """Process substring for tokenizing prefixes, infixes, suffixes and exceptions.
+
+        Args:
+            substring: The substring to tokenize.
+            start_pos: A pointer to the start position of the substring in the text.
+
+        Returns:    
+            substring: The substring to tokenize.
+            start_pos: A pointer to the start position of the substring in the text.
+            affixes: Dict holding TokenMeta lists of each affix 
+                types as result of splitting affixes
+            exception_tokens: The list of exception tokens TokenMeta objects.
+        """
+        suffixes = []
+        prefixes = []
+        infixes = []
+        exception_tokens = []
+        pos = start_pos
+        end_pos = pos
+
+        next_affix = ["prefix", "suffix"]
+
+        # Dict holding TokenMeta lists of each affix types(prefix, suffix, infix)
+        affixes = defaultdict(list)
+
+        # The first element in the `next_affix` list is 'prefix'.
+        # since we should start by finding prefixes, we fix the
+        # index i = 0.
+        # Start by finding prefixes.
+        i = 0
+
+        # Loop until an empty substring is encountered
+        while self.find_prefix(substring) or self.find_suffix(substring):
```

Comment:
The issue here is that you call the function just for the while loop. It is an overkill, calling a function, parsing the regex for suffix and prefix, just to check out if we should loop again.

The solution I suggested won't work actually:

```
while substring:
```

Because sometimes you need to stop the loop even when the substring is not empty. I agree with you on this.

So The solution in my opinion is :

to add:

```
last_substring = ""

while last_substring != substring:
    
    last_substring = substring

    # your same code here
```

Reply:
Yup, Its a better definitely